sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileClient.java

```diff
@@ -2058,11 +2070,99 @@ public ShareFileUploadInfo upload(InputStream data, long length, ParallelTransfe
      * @return The {@link ShareFileUploadInfo file upload info}
      */
     public Response<ShareFileUploadInfo> uploadWithResponse(ShareFileUploadOptions options,
-        Duration timeout, Context context) {
-        return StorageImplUtils.blockWithOptionalTimeout(
-            shareFileAsyncClient.uploadWithResponse(options, context), timeout);
+                                                            Duration timeout, Context context) {
+        StorageImplUtils.assertNotNull("options", options);
+        ShareRequestConditions validatedRequestConditions = options.getRequestConditions() == null
+            ? new ShareRequestConditions()
+            : options.getRequestConditions();
+        final ParallelTransferOptions validatedParallelTransferOptions =
+            ModelHelper.populateAndApplyDefaults(options.getParallelTransferOptions());
+        long validatedOffset = options.getOffset() == null ? 0 : options.getOffset();
+
+        final long blockSize = Math.min(validatedParallelTransferOptions.getBlockSizeLong(), 4L * 1024L * 1024L); // 4 MB
+        final int maxConcurrency = validatedParallelTransferOptions.getMaxConcurrency();
+
+        ExecutorService executor = Executors.newFixedThreadPool(maxConcurrency);
```

Comment:
Should look into designing this differently, as this will create a new thread pool per API call and will result in many, many `Thread`s being created and torn down. I'd want to look into leveraging `SharedExecutorService` in `azure-core` as the thread pool that manages parallelization here, but that generally will have many more threads available than parallelization will be set to, so we'll need to implement some level of rate limiting.

I think we should look initially into limiting the number of calls in parallel by using a [Semaphore](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Semaphore.html#Semaphore-int-) with a number of permits equal to parallelization (and remembering to release them when that upload chunk is complete).